### PR TITLE
fix for rm -rf failing when the path contains ./

### PR DESCRIPTION
**Pull request by majurgens**

### CODE REVIEW

File: rsnapshot-program.pl [hunk 1]
```diff
@@ -5254,6 +5254,19 @@ sub rm_rf {
 	# make sure we were passed an argument
 	if (!defined($path)) { return (0); }
 
+	# added by Matthew Jurgens as part of the fix for rm -rf failing when the path contains ./
+	# make sure $path does not contain ./ at the end since this makes rm -rf fail
+	# however the side effect of doing this is that we also remove the directory so we will have to create it later
+	# we could have added a * at the end so we had something like rm -rf PATH/* but probably safer not to use the *
+	$path=~s/\.\/$//;
```
Comment (DrHyde) [inline rsnapshot-program.pl:5261]: What are the circumstances in which this errant `./` appears? Better to fix that maybe?

This substitution will change a path such as `.../foo/bar./`. Should it be something like `s{/\./?$}{}` instead?

**Comment by majurgens** [comment]
I have a somewhat large answer.

In answer to "what circumstances":

The fix is for a failure scenario. It is not for any normal operation of rsnapshot.
If rsync is killled while performing a backup then you get the following:

```
rsync error: received SIGINT, SIGTERM, or SIGHUP (code 20) at rsync.c(644) [generator=3.1.3]
rsync: [generator] write error: Broken pipe (32)
rsync error: received SIGUSR1 (code 19) at main.c(1440) [receiver=3.1.3]
----------------------------------------------------------------------------
rsnapshot encountered an error! The program was invoked with these options:
/usr/bin/rsnapshot -c /tmp/rsnapshot.tmp/3.4174512.rsnapshot.conf hourly 
----------------------------------------------------------------------------
ERROR: /usr/bin/rsync returned 20 while processing user1@testhost-lan:/home/user2 /opt /etc /var/spool/cron/ /var/log/
/usr/bin/logger -p user.err -t rsnapshot[4174624] /usr/bin/rsnapshot -c \
    /tmp/rsnapshot.tmp/3.4174512.rsnapshot.conf hourly: ERROR: \
    /usr/bin/rsync returned 20 while processing user1@testhost-lan:/home/user2 /opt \
    /etc /var/spool/cron/ /var/log/ 
WARNING: Rolling back "./"
/usr/bin/logger -p user.err -t rsnapshot[4174624] WARNING: Rolling back \
    "./" 
/bin/rm -rf /data/backups/testhost-lan/hourly.0/./ 
/bin/rm: refusing to remove '.' or '..' directory: skipping '/data/backups/testhost-lan/hourly.0/./'
ERROR: Warning! /bin/rm failed.
ERROR: Error! rm_rf("/data/backups/testhost-lan/hourly.0/./")
/usr/bin/logger -p user.err -t rsnapshot[4174624] /usr/bin/rsnapshot -c \
    /tmp/rsnapshot.tmp/3.4174512.rsnapshot.conf hourly: ERROR: Error! \
    rm_rf("/data/backups/testhost-lan/hourly.0/./") 
rm -f /var/run/rsnapshot._export_shared_backups_testhost-lan.pid 

```
In this case, the `rm -rf` fails and rsnapshot leaves the partial hourly.0 behind. The next backup then moves the partial hourly.0 to hourly.1 and then has to copy all the missing files from the backup source and they cannot be linked to the hourly.1 equivalents (since they do not exist). This eats more disk space than it should.

I don't know if the `rm -rf` fails because of something specific on my system but I think it is a standard feature of bash to not delete `./` when you are in that directory. It is possible that `rm -rf ./` did not used to fail when rsnapshot was first developed or when the recovery code was written or that it does not fail on some other systems.

With regards to the rsnapshot code, this fails because the list of rollback dirs collected by rsnapshot includes `./` which is pushed on the `@rollback_points` stack by `push(@rollback_points, $$bp_ref{'dest'});` when the rsync fails. I could stop the `./` at this point but then the array entry would be empty and makes it a little more tricky for the downstream code to work out if it was really supposed to remove something or if this is an error condition. I decided that it was better to leave a rollback entry and clean it up where it was going to cause an issue.

With my fix, when rsync is killed you get:

```
rsync error: received SIGINT, SIGTERM, or SIGHUP (code 20) at rsync.c(644) [generator=3.1.3]
rsync: [generator] write error: Broken pipe (32)
rsync error: received SIGUSR1 (code 19) at main.c(1440) [receiver=3.1.3]
----------------------------------------------------------------------------
rsnapshot encountered an error! The program was invoked with these options:
/usr/bin/rsnapshot -c /tmp/rsnapshot.tmp/3.4178193.rsnapshot.conf hourly 
----------------------------------------------------------------------------
ERROR: /usr/bin/rsync returned 20 while processing user1@testhost-lan:/home/user2 /opt /etc /var/spool/cron/ /var/log/
/usr/bin/logger -p user.err -t rsnapshot[4178312] /usr/bin/rsnapshot -c \
    /tmp/rsnapshot.tmp/3.4178193.rsnapshot.conf hourly: ERROR: \
    /usr/bin/rsync returned 20 while processing user1@testhost-lan:/home/user2 /opt \
    /etc /var/spool/cron/ /var/log/ 
WARNING: Rolling back "./"
/usr/bin/logger -p user.err -t rsnapshot[4178312] WARNING: Rolling back \
    "./" 
/bin/rm -rf /data/backups/testhost-lan/hourly.0/./ 
/bin/cp -al /data/backups/testhost-lan/hourly.1/. \
    /data/backups/testhost-lan/hourly.0/. 
touch /data/backups/testhost-lan/hourly.0/ 
rm -f /var/run/rsnapshot._export_shared_backups_testhost-lan.pid 
/usr/bin/logger -p user.err -t rsnapshot[4178312] /usr/bin/rsnapshot -c \
    /tmp/rsnapshot.tmp/3.4178193.rsnapshot.conf hourly: ERROR: \
    /usr/bin/rsnapshot -c /tmp/rsnapshot.tmp/3.4178193.rsnapshot.conf \
    hourly: completed, but with some errors 
```

The `rm -rf` succeeds, rsnapshot continues and repairs hourly.0 using `cp -al` and the next backup can work properly

In answer to the alternate regex:
Yes, that is superior. I will update the code to use it.
The new regex fails to convert `"./"` to "" but I think this will be ok since it should always be preceded by some other path structure by the time it gets to the `rm -rf` and it is also ok to remove the trailing `/` from a path like `/a/b/c/` when passing to the `rm -rf`.
I have tested this new regex in the failure scenario and it works ok

**Comment by sgpinkus** [comment]
So to summarize the root cause of this issue is "./"s in the backup destination path? @majurgens can you please paster your config, or preferably a simplified version with a  `backup` that triggers this bug? I want to reproduce.

**Comment by majurgens** [comment]
I think the key config items are:
```
snapshot_root   /export/test
link_dest       1
backup  /test/     ./
verbose         5
loglevel        3
```
Using ./ in the backup destination to simplify the resulting backup structure

Follow these steps:
1. Make sure /test is big enough so that when it is backed up it will take a few seconds. I find removing all previous backups helps so that the rsync has to copy everything again
2. Run the rsnapshot
3. Send the kill signal to the rsync process to simulate an error condition. This is where you need the backup to be running long enough to be still running when you kill it

After the rsync process dies, this is where rsnapshot tries to clean up after it and run the rm -rf with the path ending in ./ eg /export/test/hourly.0/./

When you reproduce you should see logs similar to what I previously pasted

**Comment by DrHyde** [comment]
Ah-ha! I wondered where the ./ was coming from and just *ass*umed it was an artifact of link_dest. So you're only backing up a single volume, and having all of its files/directories directly under $snapshot_root/daily.{0,1,2,...}?

That's an Unusual setup. Normally even if you're only backing up one volume you'd arrange things to look like this ...

```
snapshot_root/daily.0/original_volume_name/foo.txt
snapshot_root/daily.0/original_volume_name/foo.d
snapshot_root/daily.0/original_volume_name/foo.d/bar
...
```

What you're doing is a bit of a hack but I can see why you'd want to do it :-)

**Comment by majurgens** [comment]
`snapshot_root/daily.0/original_volume_name/foo.txt`

I like sorting my backups by hostname (which is part of `snapshot_root` and hence the additional path `original_volume_name` is redundant

**Comment by DrHyde** [comment]
I presume that this means you're running rsnapshot once per hostname then.

Anyway, thanks for the patch, I've merged it now.

**Comment by majurgens** [comment]
Yes, once per hostname. My backup storage structure looks like:
/backups/hostX/hourly.0/sourcePath

File: rsnapshot-program.pl [hunk 2]
```diff
@@ -5254,6 +5254,19 @@ sub rm_rf {
...
+	# in order to recreate the directory we will want to take a look at it before we remove it
```
Comment (sgpinkus) [inline rsnapshot-program.pl:5262]: This has caused a regression: with lazy deletes empty _delete.xxxx directories are left in place after successful execution. TBH I don't get why this isn't a one liner. Why do we need to restore the directory?

File: rsnapshot-program.pl [hunk 3]
```diff
@@ -5274,6 +5287,20 @@ sub rm_rf {
 		$result = rmtree("$path", 0, 0);
 	}
 
+	# added by Matthew Jurgens as part of the fix for rm -rf failing when the path contains ./
+	# make sure the directory is still in place
+	# if you cannot create this directory and this rm is part of a rollback which uses cp -al, then the cp -al will fail shortly after 
+	# MKDIR DEST (AND SET MODE)
+	if (defined($st)) {
```
Comment (sgpinkus) [inline rsnapshot-program.pl:5294]: Why?

**Pull request merged**